tests/components/picnic/test_services.py

```diff
@@ -0,0 +1,199 @@
+"""Tests for the Picnic services."""
+from functools import lru_cache
+from unittest.mock import MagicMock, patch
+
+import pytest
+
+from homeassistant.components.picnic import CONF_COUNTRY_CODE, DOMAIN
+from homeassistant.components.picnic.const import SERVICE_ADD_PRODUCT_TO_CART
+from homeassistant.components.picnic.services import PicnicServiceException
+from homeassistant.const import CONF_ACCESS_TOKEN
+from homeassistant.core import HomeAssistant
+
+from tests.common import MockConfigEntry
+
+
+@pytest.fixture
+def picnic_api_client():
+    """Create PicnicAPI mock with set response data."""
+    with patch(
+        "homeassistant.components.picnic.create_picnic_client"
+    ) as create_picnic_client_mock:
+
+        @lru_cache(maxsize=2, typed=True)
```

Comment:
Why is this needed?

Reply:
Ah, I see why. Typically the way to do this is use a separate fixture for the part you want to get, then the test can also ask for it (e.g. a "picknic_api" fixture.)

I think the reason you need a generator is because there is a test where a second config entry is created. What I would suggest is use the common fixtures for the normal case, then do something special for the case where you need to create second entry. Then you don't need this lru cache annotation, and the code gets simpler in the common test cases.

Reply:
I don't think it was worth the effort, but I've changed it!

Reply:
Thank you i appreciate your effort here and find the tests significantly easy to follow now.